File: apps/agent/src/tools/upgradeToPro.ts
import { FunctionTool } from '@google/adk';
import { z } from 'zod';

/**
 * upgrade_to_pro — UI-directive tool. Mirrors connect_workspace / auth_user.
 *
 * The LLM emits this when a heavy free user could plausibly benefit from
 * Pro (more depth, faster replies, no daily nudges). The tool response is
 * a pure UI signal the web picks up via SSE; the browser renders an
 * UpgradePrompt card. Real billing isn't wired yet — for this phase the
 * card just collects interest.
 *
 * IMPORTANT: like connect_workspace, the LLM is never involved in payment.
 * This tool has no args, returns no billing values, and must not see any
 * customer/subscription/price IDs. Same auth-plane boundary.
 */
export const UPGRADE_TO_PRO_TOOL_NAME = 'upgrade_to_pro';

export function createUpgradeToProTool(): FunctionTool {
  const parameters = z.object({});

  // biome-ignore lint/suspicious/noExplicitAny: zod nominal mismatch with ADK generics
  return new FunctionTool<any>({
    name: UPGRADE_TO_PRO_TOOL_NAME,
    description:
      'Surface a Lifecoach Pro upgrade card to the user. Use sparingly — at most once per ' +
      'session, and only when the conversation has just hit a moment where Pro would genuinely ' +
      'help (deeper analysis, faster replies, no daily nudges). After calling, write NO ' +
      'additional text that turn — the upgrade card is the entire response. Do NOT attempt to ' +
      'handle any payment values yourself; the application handles billing if and when the user ' +
      'opts in.',
    parameters,
    execute: async () => {
      return {
        status: 'upgrade_prompted' as const,
      };
    },
  });
}

File: apps/agent_py/src/lifecoach_agent/tools/upgrade_to_pro.py
"""upgrade_to_pro tool — UI-directive only.

Surfaces an UpgradePrompt card to the web. The LLM never sees billing
identifiers or values — same auth-plane boundary as connect_workspace.
"""

from __future__ import annotations

from typing import Any

UPGRADE_TO_PRO_TOOL_NAME = "upgrade_to_pro"


async def upgrade_to_pro() -> dict[str, Any]:
    """Surface a Lifecoach Pro upgrade card to the user. Use sparingly —
    at most once per session, and only when the conversation has just hit
    a moment where Pro would genuinely help (deeper analysis, faster
    replies, no daily nudges). After calling, write NO additional text
    that turn — the upgrade card is the entire response. Do NOT attempt
    to handle any payment values yourself; the application handles
    billing if and when the user opts in."""
    return {"status": "upgrade_prompted"}


def create_upgrade_to_pro_tool() -> Any:
    from google.adk.tools import FunctionTool

    return FunctionTool(upgrade_to_pro)
